internal/util/pool.go

```diff
@@ -6,105 +6,171 @@ import (
 	"sync"
 )
 
+// The result of running a map function in parallel pool.
+// Always call IsDone() before accessing any results to avoid concurrency issues.
+type PoolResult[InputType comparable, OutputType any] struct {
+	// The results of the parallel pool.
+	Results map[InputType]OutputType
+
+	// A map of work errors using the item for item-level errors.
+	WorkErrors map[InputType]error
+
+	// Signals that the parallel pool was canceled during execution.
+	Canceled bool
+
+	// An internal done channel to signal the results can be accessed.
+	done chan any
+}
+
+// Use this method to block until the results can be accessed.
+func (this PoolResult[InputType, OutputType]) IsDone() {
+	if this.done != nil {
+		<-this.done
+	}
+}
+
 // Do a map function (one result for one input) with a parallel pool of workers.
 // Unless there is a critical error (the final return value) or cancellation,
-// the output will have the same length as and index-match the input, but will have an empty value if there is an error.
-// A cancellation will return (nil, nil, nil).
-// Consult the returned map of errors using the item's index to check for item-level errors.
+// every input will be in PoolResult.Results.
```

Comment:
This seems strange.
Why are we putting errors in the results?
It seems like it should be either Results or WorkErrors.